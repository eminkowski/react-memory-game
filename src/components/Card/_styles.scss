@import '~globals';

.card {
  flex: 0 0 50px;
  height: 80px;
  margin: 10px;
  perspective: 800px;
  position: relative;
  user-select: none;

  @media (min-width: 500px) {
    flex: 0 0 90px;
    height: 120px;
  }

  @media (min-width: 1000px) {
    flex: 0 0 130px;
    height: 160px;
  }

  &-container {
    height: 100%;
    position: absolute;
    transform-style: preserve-3d;
    transition: transform .3s;
    width: 100%;

    &--hidden {
      transform: rotateY(180deg);
  	}
  }

  &-front,
  &-back {
    backface-visibility: hidden;
    color: $black;
    display: block;
    font-size: 30px;
    font-weight: 600;
    height: 100%;
    line-height: 80px;
    margin: 0;
    position: absolute;
    text-align: center;
    width: 100%;

    @media (min-width: 500px) {
      font-size: 40px;
      line-height: 120px;
    }

    @media (min-width: 1000px) {
      font-size: 50px;
      line-height: 160px;
    }
	}

  &-front {
    background-color: $white;
	}

  &-back {
    background: linear-gradient(to bottom, $purple, $purple 50%, $pink 50%, $pink);
    background-size: 100% 20px;
    cursor: pointer;
    transform: rotateY(180deg);
	}

  &--matched {
    background-color: transparent;
    color: transparent;
    transition: all 1.5s ease .5s;
  }
}
