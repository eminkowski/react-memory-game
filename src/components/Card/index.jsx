import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';

import './_styles.scss';

export default class Card extends PureComponent {
  static propTypes = {
    className: PropTypes.string,
    style: PropTypes.string,
    id: PropTypes.string.isRequired,
    value: PropTypes.string.isRequired,
    isFlipped: PropTypes.bool.isRequired,
    isMatch: PropTypes.bool.isRequired,
    onClick: PropTypes.func.isRequired,
  };
  
  handleClick = (e) => {
    const { id, value, isFlipped, onClick } = this.props;
    
    // Do not proceed unless the card has been flipped
    if (!isFlipped) {
      // Call parent function with value and id
      onClick(value, id);
    }
  }
  
  render() {
    const { className, style, value, isFlipped, isMatch } = this.props;

    // Conditionally add styling for flipped and matched
    return (
      <section
        className={cn('card', className)}
        style={style}
        onClick={this.handleClick}
      >
        <div className={cn('card-container', { 'card-container--hidden': !isFlipped })}>
          <figure className={cn('card-front', { 'card--matched': isMatch })}>
            {value}
          </figure>
          <figure className="card-back" />
        </div>
      </section>
    );
  }
};
