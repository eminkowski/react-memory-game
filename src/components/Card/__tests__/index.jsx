/* eslint-env jest */
import React from 'react';
import renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';

import Card from '../index';

describe('Card', () => {
  it('should render correctly', () => {
    const component = renderer.create(
      <Card id="1" value="1" isFlipped isMatch onClick={jest.fn()} />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should display the card value that matches its prop value of 1', () => {
    const component = shallow(
      <Card id="1" value="1" isFlipped isMatch onClick={jest.fn()} />
    );
    const front = component.find('.card-front');
    expect(front.text()).toEqual('1');
  });

  it('should have the hidden class when the card is not flipped', () => {
    const component = shallow(
      <Card id="1" value="1" isFlipped={false} isMatch={false} onClick={jest.fn()} />
    );
    const container = component.find('.card-container');
    expect(container.hasClass('card-container--hidden')).toEqual(true);
  });

  it('should -not- have the hidden class when the card is flipped', () => {
    const component = shallow(
      <Card id="1" value="1" isFlipped isMatch={false} onClick={jest.fn()} />
    );
    const container = component.find('.card-container');
    expect(container.hasClass('card-container--hidden')).toEqual(false);
  });

  it('should -not- have the matched class when the card has no match', () => {
    const component = shallow(
      <Card id="1" value="1" isFlipped isMatch={false} onClick={jest.fn()} />
    );
    const front = component.find('.card-front');
    expect(front.hasClass('card--matched')).toEqual(false);
  });

  it('should have the matched class when the card has a match', () => {
    const component = shallow(
      <Card id="1" value="1" isFlipped isMatch onClick={jest.fn()} />
    );
    const front = component.find('.card-front');
    expect(front.hasClass('card--matched')).toEqual(true);
  });

  it('should render a figure with the onClick prop', () => {
    const onClick = jest.fn();

    const component = mount(
      <Card id="1" value="1" isFlipped={false} isMatch onClick={onClick} />
    );
    const figure = component.get(0);
    expect(figure.props.onClick).toBe(onClick);
  });
});
