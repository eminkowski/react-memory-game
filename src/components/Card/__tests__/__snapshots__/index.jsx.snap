// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card should render correctly 1`] = `
<section
  className="card"
  onClick={[Function]}
  style={undefined}
>
  <div
    className="card-container"
  >
    <figure
      className="card-front card--matched"
    >
      1
    </figure>
    <figure
      className="card-back"
    />
  </div>
</section>
`;
